src/Statistics.jl

```diff
@@ -0,0 +1,115 @@
+using Random: shuffle!
+@doc doc"""
+    mean(M,x; weights=1/n*ones(n), stop_iter=100, kwargs... )
+
+computes the Riemannian center of mass also known as Karcher mean of the vector
+`x` of `n` points on the [`Manifold`](@ref) `M`. This function
+uses the gradient descent scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian center of mass.
+The general formula to compute the minimizer reads
+````math
+\argmin_{y\in\mathcal M} \frac{1}{2}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}^2(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref).
+
+Finally `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the inimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in 
+> Afsari, B; Tron, R.; Vidal, R.: On the Convergence of Gradient
+> Descent for Finding the Riemannian Center of Mass,
+> SIAM Journal on Control and Optimization (2013), 51(3), pp. 2230–2260, 
+> doi: [10.1137/12086282X](https://doi.org/10.1137/12086282X),
+> arxiv: [1201.0925](https://arxiv.org/abs/1201.0925">1201.0925)
+"""
+function mean(M::Manifold, x::AbstractVector; kwargs...)
+    y = x[1]
```

Comment:
I'd use `similar_result(M, mean, x[1], x)` here.

Reply:
Yes, no problem.

Reply:
Hm, no, there is a problem: The function is not defined for these types of input?

Reply:
Right, this won't work. `similar_result(M, mean, x[1])` should be good enough.

Reply:
Hm, throws the same error message (no method matching) and I am not an expert on the idea of `similar_result` here.

Reply:
I see now, that's a bug. `Manifolds.jl` doesn't import `similar_result` from `ManifoldsBase.jl` but it should. I'm really surprised this didn't cause any other problems.

Reply:
Interesting – I fixed it but still have to set `y` to `x[1]` then, because `similar` returns zero values in all entries.

Reply:
Yes, since `mean!` is supposed to use `y` as a starting point, you should use `copy(x[1])` here (not just `x[1]` since it will be overwritten).

Reply:
Still, I have some side-effect somewhere, at least in the median; despite that your proposed improvements seem to be very good :)